File: medium/6_Zigzag_Conversion/main.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        int len = s.size();

        if(len <= numRows || numRows == 1){
            return s;
        }
        
        vector<string> strRow (numRows,"");

        int i = 0, j;
        while(i < len){
            for(j = 0; j < numRows && i < len; j++){
                strRow[j] += s[i];
                i++;
            }

            for(j = numRows - 2; j > 0 && i < len; --j){
                strRow[j] += s[i];
                i++;
            }
        }

        string res = "";

        for(int i = 0; i < numRows; i++){
            res += strRow[i];
        }

        return res;
    }
};

File: medium/6_Zigzag_Conversion/main.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        length = len(s)

        if length <= numRows or numRows == 1:
            return s

        rows = [[] for _ in range(numRows)]
        
        i = 0
        step = -1
        for char in s:
            rows[i].append(char)
            if i == 0:
                step = 1
            elif i == numRows - 1:
                step = -1
            i += step

        for i in range(numRows):
            rows[i] = ''.join(rows[i])
        return ''.join(rows)         